Added event for temporary lockout for brute force protector

tsaarni opened the pull request:
This PR adds new event when user account is temporarily disabled by brute force protector. Before this change, and event was sent for only permanent lockout. Having event also for temporary lockout can improve visibility of ongoing brute force attack.

The PR also removes the free-text log for failed login attempts. This avoids having following printout in logs:

```
2024-01-31 10:51:41,466 WARN  [org.keycloak.services] (Brute Force Protector) KC-SERVICES0053: login failure for user 69563c18-7215-471c-b344-c7a0a8d49296 from ip 127.0.0.1
```

Reasoning for this change is that unconditionally printing user id and IP address in free-text logs for every invalid login may be seen as unnecessary logging of personally identifiable information. This can be also better served by event now that both permanent and temporary lockout are covered. In lockout event the presence of this information is justified and due to the structured format of events, is easier to process appropriately.  

Fixes #26626

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

tsaarni committed:
Event for temporary lockout brute force protector

This change adds event for brute force protector when user account is
temporarily disabled.

It also lowers the priority of free-text log for failed login attempts.

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

tsaarni commented:
One observation regarding the default implementation `JBossLoggingEventListenerProvider.java`: 

It converts "success" events to `DEBUG` level logs by default  ([code here](https://github.com/keycloak/keycloak/blob/b41e2f82c4f8bcf9c0e35e15119ba6ad89c3ccba/services/src/main/java/org/keycloak/events/log/JBossLoggingEventListenerProvider.java#L84)).  Since by default the [log level](https://www.keycloak.org/server/logging) is set to `INFO`, neither permanent or temporary lockout is visible if using  `jboss-logging` event listener. This can be changed by using `--spi-events-listener-jboss-logging-success-level` or env `KC_SPI_EVENTS_LISTENER_JBOSS_LOGGING_SUCCESS_LEVEL` ([doc here](https://www.keycloak.org/server/all-provider-config#_jboss_logging)) and setting log level that will be logged by default, for example `INFO`.

ahus1 commented on services/src/main/java/org/keycloak/services/ServicesLogger.java:
```diff
@@ -253,7 +253,7 @@ public interface ServicesLogger extends BasicLogger {
     void failedProcessingType(@Cause Exception e);
 
-    @LogMessage(level = WARN)
+    @LogMessage(level = DEBUG)
```
I agree to change this to DEBUG, and also to eventually remove it in a future release as the `LOGIN_ERROR` event which is also logged contains the same information. 

IMHO you could go as far as removing it. Either way, it needs to be mentioned in the release notes, as people might depend on this log message, and should recommend to use the `LOGIN_ERROR` event instead.

tsaarni replied:
I've removed the log message completely. 

Now I must confess, I'm not aware how I can add a note to the release notes?

ahus1 replied:
I'll prepare a suggestion and push it to this branch. Then you can review it.

ahus1 reviewed (changes requested):
Thank you for this change, I like the direction. See below for some comments that I'd like to see changed.

tsaarni committed:
Remove brute force protector log for invalid login

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

tsaarni committed:
Add not_before and num_failures to brute force events

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

ahus1 commented:
@tsaarni - I've seen that you have merged the main branch into this PR. Within the Keycloak project, we usually prefer a rebase, as that allows for simpler squashing of commits later.

tsaarni commented:
> @tsaarni - I've seen that you have merged the main branch into this PR. Within the Keycloak project, we usually prefer a rebase, as that allows for simpler squashing of commits later.

Sure! I've removed the merge commit, rebased and force pushed - also to re-trigger CI since there seemed to be some problems with Quarkus 3.7.0.CR1 that I think were unrelated.

tsaarni commented on services/src/main/java/org/keycloak/services/managers/DefaultBruteForceProtector.java:
```diff
@@ -191,6 +187,7 @@ protected void failure(KeycloakSession session, LoginEvent event) {
                 int notBefore = (int) (currentTime / 1000) + waitSeconds;
                 logger.debugv("set notBefore: {0}", notBefore);
                 userLoginFailure.setFailedLoginNotBefore(notBefore);
```
Good idea! I've added new event detail attributes for these.

ahus1 replied:
Thanks, this resolves this comment for me.

tsaarni commented on services/src/main/java/org/keycloak/services/managers/DefaultBruteForceProtector.java:
```diff
@@ -168,20 +172,15 @@ protected void failure(KeycloakSession session, LoginEvent event) {
             userLoginFailure.incrementFailures();
             logger.debugv("new num failures: {0}", userLoginFailure.getNumFailures());
 
-            if(userLoginFailure.getNumFailures() == realm.getFailureFactor()) {
+            if(userLoginFailure.getNumFailures() >= realm.getFailureFactor()) {
```
I noticed a tiny bug, which I fixed here: 

when changing brute force protector settings meanwhile brute force protector was already active, it could happen that number of failures got greater than failure factor and `==` was never true.

ahus1 replied:
Good point. Still I'm wondering if this condition might now need addition clauses as we want it to fire only once per lockout. So one might need to check that the user is either enabled or the disabled reason is not set to `DISABLED_BY_PERMANENT_LOCKOUT`. I have to admit I'm only looking at the local code and not at testing this in my IDE.

tsaarni replied:
I will test this under debugger.

tsaarni replied:
It seems will not be another call anymore to `DefaultBruteForceProtector.failure()` method after the user permanently is locked, and therefore the condition is never checked again. 

The processing of login is stopped already earlier, during the authentication flows when validating the user, for example here

https://github.com/keycloak/keycloak/blob/50afa72266a8a1c8b20413912564718f096c4209/services/src/main/java/org/keycloak/authentication/authenticators/browser/AbstractUsernameFormAuthenticator.java#L228

or here

https://github.com/keycloak/keycloak/blob/50afa72266a8a1c8b20413912564718f096c4209/services/src/main/java/org/keycloak/authentication/authenticators/directgrant/ValidateUsername.java#L80

tsaarni replied:
Even though the code does not seem to be called, if you think it would be safer I could still add a new condition, for example like this

```diff
@@ -174,7 +173,7 @@ public class DefaultBruteForceProtector implements Runnable, BruteForceProtector

             if(userLoginFailure.getNumFailures() >= realm.getFailureFactor()) {
                 UserModel user = session.users().getUserById(realm, userId);
-                if (user == null) {
+                if (user == null || user.isEnabled() == false) {
                     return;
                 }
                 logger.debugv("user {0} locked permanently due to too many login attempts", user.getUsername());
```

pedroigor replied:
Does it make more sense to be consistent with the original scope of this PR and avoid fixing the brute-force logic?

tsaarni replied:
Sure, I removed the fix for the comparison of number of failures.

pedroigor reviewed (dismissed):
@tsaarni Thanks for your time on it.

ahus1 commented:
I've submitted a PR with documentation changes to this PR for @tsaarni to review and merge: https://github.com/Nordix/keycloak/pull/235

ahus1 committed:
Updating documentation including release and migration notes (#235)

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

tsaarni commented:
> I've submitted a PR with documentation changes to this PR for @tsaarni to review and merge: [Nordix#235](https://github.com/Nordix/keycloak/pull/235)

Thanks @ahus1! 

DCO check seems unhappy, maybe your commit should not have had my sign-off.  Do you want it fixed?

jonkoops commented:
I've set the DCO to pass, since @ahus1 is a member of the Keycloak team this should not be an issue.

ahus1 reviewed (approved):
Approving based on previous reviews of @pedroigor and the docs review by @tsaarni.
